mantidimaging/core/io/saver.py

```diff
@@ -178,7 +178,7 @@ def nexus_save(dataset: StrictDataset, path: str, sample_name: str):
         with h5py.File(path, "w", driver="core") as nexus_file:
             _nexus_save(nexus_file, dataset, sample_name)
     except OSError:
-        pass
+        os.remove(path)
```

Comment:
For some error cases there might not have been anything written so far, so the `os.remove()` will raise an `OSError` as well.

For example if `path` is somewhere read-only (like `/lib/a.nxs` ), then I get:
```
Traceback (most recent call last):
  File "/home/sam/git/mantidimaging/mantidimaging/gui/windows/main/presenter.py", line 81, in notify
    self.save_nexus_file()
  File "/home/sam/git/mantidimaging/mantidimaging/gui/windows/main/presenter.py", line 138, in save_nexus_file
    self.model.do_nexus_saving(dataset_id, self.view.nexus_save_dialog.save_path(),
  File "/home/sam/git/mantidimaging/mantidimaging/gui/windows/main/model.py", line 90, in do_nexus_saving
    saver.nexus_save(self.datasets[dataset_id], path, sample_name)  # type: ignore
  File "/home/sam/git/mantidimaging/mantidimaging/core/io/saver.py", line 181, in nexus_save
    os.remove(path)
FileNotFoundError: [Errno 2] No such file or directory: '/lib/aaa'
2022-04-21 08:49:44,521 [mantidimaging.gui.windows.main.presenter:L89] ERROR: Notification handler failed
Traceback (most recent call last):
  File "/home/sam/git/mantidimaging/mantidimaging/core/io/saver.py", line 178, in nexus_save
    with h5py.File(path, "w", driver="core") as nexus_file:
  File "/home/sam/miniconda3/envs/mantidimaging-dev/lib/python3.8/site-packages/h5py/_hl/files.py", line 455, in __init__
    fid = make_fid(name, mode, userblock_size,
  File "/home/sam/miniconda3/envs/mantidimaging-dev/lib/python3.8/site-packages/h5py/_hl/files.py", line 205, in make_fid
    fid = h5f.create(name, h5f.ACC_TRUNC, fapl=fapl, fcpl=fcpl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5f.pyx", line 120, in h5py.h5f.create
OSError: Unable to create file (unable to open file)
```

I don't think checking if the file exists before removing helps, because  if `path` is an existing read only file. The write could fail, but the delete could succeed.
```
mkdir -p /tmp/nexus_test
echo "something" > /tmp/nexus_test/existing_file.nxs
chmod -w /tmp/nexus_test/existing_file.nxs
```
If I try to save to `/tmp/nexus_test/existing_file.nxs`, then that file gets deleted.

So might be best to have `h5py.File(path, "w", driver="core")` and `_nexus_save(nexus_file, dataset, sample_name)` in different try blocks. So the remove only happens if teh exception is in `_nexus_save`